<a href="https://colab.research.google.com/github/itskevalchaudhari/AIML-Practicals/blob/main/Practical-09/Practical-09-Object-Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practical 09

## Aim

To perform object detection on CCTV footage using a pre-trained YOLO (You Only Look Once) deep learning model and identify objects present in the video using bounding boxes and class labels.

In [ ]:
# Practical 09
# Object Detection from CCTV Footage

print("Practical 09: Object Detection from CCTV Footage")

## Theory

Object detection is a computer vision technique used to identify and locate objects in an image or video. Unlike image classification, object detection provides both the **class label** of an object and its **location** using a bounding box.

CCTV footage consists of a sequence of video frames. Object detection can be applied to each frame to identify objects such as people, cars, bicycles, buses, and other objects.

### YOLO (You Only Look Once)

YOLO is a real-time object detection algorithm that performs object detection in a single neural network evaluation. It processes the complete image at once and predicts bounding boxes and class probabilities simultaneously.

In this practical, a **pre-trained YOLO model** is used to detect objects in CCTV footage.

### Main Components of YOLO

1. **Bounding Box** – Represents the location of a detected object.
2. **Class Label** – Identifies the detected object.
3. **Confidence Score** – Indicates how confident the model is about the detection.
4. **Non-Maximum Suppression (NMS)** – Removes overlapping duplicate bounding boxes.

### Bounding Box Parameters

A detected bounding box can be represented using:

- x = center x-coordinate
- y = center y-coordinate
- w = width of the bounding box
- h = height of the bounding box

The bounding box is commonly represented as:

**(x, y, w, h)**

### Intersection over Union (IoU)

IoU measures the overlap between the predicted bounding box and the actual bounding box.

**IoU = Area of Intersection / Area of Union**

A higher IoU indicates better overlap between the two bounding boxes.

### Confidence Score

The confidence score represents the model's confidence that a detected bounding box contains an object of a particular class.

Object detection can be used in CCTV surveillance for applications such as people counting, vehicle detection, traffic monitoring, security monitoring, and suspicious activity analysis.

In [1]:
# Install Ultralytics YOLO

!pip install -q ultralytics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.9/46.9 kB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 26.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.1/58.1 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.5/96.5 kB 6.4 MB/s eta 0:00:00


In [2]:
# Import required libraries

from ultralytics import YOLO
import cv2
import matplotlib.pyplot as plt

print("Ultralytics YOLO imported successfully.")

Creating new Ultralytics Settings v0.0.8 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/usage/settings.
Ultralytics YOLO imported successfully.


In [3]:
# Load the pre-trained YOLO model

model = YOLO("yolo11n.pt")

print("YOLO model loaded successfully.")

YOLO model loaded successfully.


In [5]:
# Upload CCTV video

from google.colab import files

uploaded = files.upload()

Saving Felony_Hit-and-Run.webm to Felony_Hit-and-Run.webm


In [6]:
# Perform object detection on CCTV footage

input_video = "Felony_Hit-and-Run.webm"

results = model.predict(
    source=input_video,
    conf=0.25,
    save=True
)

print("Object detection completed successfully.")


WARNING ⚠️ 
Inference results will accumulate in RAM unless `stream=True` is passed, which can cause out-of-memory errors for large
sources or long-running streams and videos. See https://docs.ultralytics.com/modes/predict for help.

Example:
    results = model(source=..., stream=True)  # generator of Results objects
    for r in results:
        boxes = r.boxes  # Boxes object for bbox outputs
        masks = r.masks  # Masks object for segment masks outputs
        probs = r.probs  # Class probabilities for classification outputs

video 1/1 (frame 1/202) /content/Felony_Hit-and-Run.webm: 384x640 1 person, 3 cars, 451.2ms
video 1/1 (frame 2/202) /content/Felony_Hit-and-Run.webm: 384x640 1 person, 3 cars, 125.8ms
video 1/1 (frame 3/202) /content/Felony_Hit-and-Run.webm: 384x640 1 person, 3 cars, 118.0ms
video 1/1 (frame 4/202) /content/Felony_Hit-and-Run.webm: 384x640 1 person, 3 cars, 116.8ms
video 1/1 (frame 5/202) /content/Felony_Hit-and-Run.webm: 384x640 1 person, 3 cars, 124.3ms

In [8]:
# Find the processed CCTV video

import os
import glob

video_files = []

for root, dirs, files in os.walk("runs"):
    for file in files:
        if file.lower().endswith((".mp4", ".avi", ".mov", ".webm", ".mkv")):
            video_files.append(os.path.join(root, file))

print("Processed video files found:")

for file in video_files:
    print(file)

Processed video files found:
runs/detect/predict/Felony_Hit-and-Run.avi


In [ ]:
# Display the processed CCTV video

from IPython.display import Video, display

output_video = "runs/detect/predict/Felony_Hit-and-Run.avi"

print("Processed video:", output_video)

display(Video(output_video, embed=True))

In [10]:
# Convert the detected AVI video to MP4

import subprocess

input_video = "runs/detect/predict/Felony_Hit-and-Run.avi"
output_video = "runs/detect/predict/Felony_Hit-and-Run_Detected.mp4"

subprocess.run([
    "ffmpeg",
    "-y",
    "-i", input_video,
    "-c:v", "libx264",
    "-pix_fmt", "yuv420p",
    output_video
], check=True)

print("Detected video converted to MP4 successfully.")
print("Output:", output_video)

Detected video converted to MP4 successfully.
Output: runs/detect/predict/Felony_Hit-and-Run_Detected.mp4


In [ ]:
# Display the final detected CCTV video

from IPython.display import Video, display

final_video = "runs/detect/predict/Felony_Hit-and-Run_Detected.mp4"

print("Final detected video:")
print(final_video)

display(Video(final_video, embed=True))

In [12]:
# Object Detection Summary

from collections import Counter

detected_objects = []

for result in results:
    if result.boxes is not None:
        for cls in result.boxes.cls:
            detected_objects.append(model.names[int(cls)])

object_counts = Counter(detected_objects)

print("Object Detection Summary")
print("------------------------")

for obj, count in object_counts.items():
    print(f"{obj}: {count} detections")

print("\nTotal detections:", len(detected_objects))

Object Detection Summary
------------------------
car: 716 detections
person: 87 detections
bus: 3 detections

Total detections: 806


## Conclusion

In this practical, object detection was successfully performed on CCTV footage using a pre-trained YOLO model.

The model processed the video frame-by-frame and detected multiple objects using bounding boxes, class labels, and confidence scores. The detected objects included cars, persons, and buses.

A total of **806 object detections** were obtained from the processed CCTV footage, including **716 cars, 87 persons, and 3 buses**.

The practical demonstrates that YOLO can be effectively used for real-time object detection and CCTV surveillance applications such as traffic monitoring, security monitoring, and vehicle detection.

## Result

Object detection was successfully performed on CCTV footage using the YOLO deep learning model. The detected objects were identified with bounding boxes, class labels, and confidence scores, demonstrating the practical application of deep learning in video-based surveillance.